### PR TITLE
Improve tests

### PR DESCRIPTION
**Pull request by sjaeckel**
a.k.a I found an error that I couldn't localize yet ... ~~and now I hope @czurnieden is quicker to find out why `mp_prime_is_prime()` doesn't work with the stripped-down version...~~ found and fixed it!

### CODE REVIEW

**Comment by sjaeckel** [comment]
okay, I've found out that it's caused by the way the perl script creates `tommath_class.h` and misses the function pointers to the different reduce functions (the ones assigned to `redux`) in bn_s_mp_exptmod.c resp. bn_s_mp_exptmod_fast.c

The following hack + rerun `helper.pl -u` fixes this error...
```
diff --git a/bn_s_mp_exptmod_fast.c b/bn_s_mp_exptmod_fast.c
index 56e2b2b..fcee498 100644
--- a/bn_s_mp_exptmod_fast.c
+++ b/bn_s_mp_exptmod_fast.c
@@ -94,2 +94,6 @@ mp_err s_mp_exptmod_fast(const mp_int *G, const mp_int *X, const mp_int *P, mp_i
          goto LBL_M;
+         s_mp_montgomery_reduce_fast(NULL, NULL, 0);
+         mp_montgomery_reduce(NULL, NULL, 0);
+         mp_dr_reduce(NULL, NULL, 0);
+         mp_reduce_2k(NULL, NULL, 0);
       }
```

File: demo/test.c [hunk 1]
```diff
@@ -1543,262 +1705,277 @@ static int test_mp_ilogb(void)
    int size;
    const uint32_t max_base = MP_MIN(UINT32_MAX, MP_DIGIT_MAX);
```
Comment (czurnieden) [inline demo/test.c]: The conversion check complaint here (https://travis-ci.org/libtom/libtommath/jobs/581756157)
(No, I don't know how the extra forward slash came into existence at that place)
My error?

**Comment by minad** [comment]
Hmm, I am not sure if I like those changes. Instead of adding MP_HAS checks to every test (a lot of boilerplate), just check for MP_HAS in the test registry and only add the functions to the table which are enabled. The other test__x functions won't be linked in.

**Comment by minad** [comment]
~~~
registry [] = { MP_HAS(X) ? test_x : 0, ... }
~~~

File: demo/test.c [hunk 2]
```diff
@@ -2219,54 +2224,61 @@ int unit_tests(int argc, char **argv)
...
-      T(mp_prime_rand),
-      T(mp_rand),
-      T(mp_read_radix),
-      T(mp_reduce_2k),
-      T(mp_reduce_2k_l),
+#define T_OFF           {"", NULL}
+#define T(n, o)         { #n, MP_HAS(o) ? test_##n : NULL }
+#define T2(n, o1, o2)   { #n, MP_HAS(o1) && MP_HAS(o2) ? test_##n : NULL }
+      T(feature_detection, TEST_FEATURE_DETECTION),
+      T(trivial_stuff, TEST_TRIVIAL_STUFF),
```
Comment (minad) [inline demo/test.c]: Instead define T0, T1 and T2? Besides that it looks good.

Comment (sjaeckel) [reply demo/test.c]: done!
now I've only to find out why that single build fails...

File: demo/test.c [hunk 3]
```diff
@@ -59,6 +60,8 @@ static uint64_t uabs64(int64_t x)
  */
 void does_not_exist(void);
 
+#define BN_TEST_FEATURE_DETECTION_C
```
Comment (minad) [inline demo/test.c]: If you define T0 you can avoid these unnecessary defines

File: demo/test.c [hunk 4]
```diff
@@ -93,6 +96,8 @@ static int test_feature_detection(void)
    return EXIT_SUCCESS;
 }
 
+#define BN_TEST_TRIVIAL_STUFF_C
+
```
Comment (minad) [inline demo/test.c]: Same here

File: etc/makefile [hunk 5]
```diff
@@ -1,4 +1,5 @@
-CFLAGS += -Wall -W -Wextra -Wshadow -O3 -I../
+LTM_CFLAGS += -Wall -W -Wextra -Wshadow -O3 -I../
```
Comment (minad) [inline etc/makefile]: Why complicating things here? the user could already add flags to CFLAGS, couldn't they?

Comment (sjaeckel) [reply etc/makefile]: yeah but only through `CFLAGS="-foo" make` now it's also possible via `make CFLAGS="-foo"`

Comment (minad) [reply etc/makefile]: Ok

**Pull request merged**